sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -1102,13 +1097,56 @@ def get_cmis_dp_init_duration_secs(self, api):
     def get_cmis_dp_deinit_duration_secs(self, api):
         return api.get_datapath_deinit_duration()/1000
 
-    def is_cmis_application_update_required(self, api, channel, speed):
+    def get_cmis_host_lane_mask(self, api, host_lane_count, channel, speed):
+        """
+        Retrieves mask of active host lanes based on host lane count, channel and speed
+
+        Args:
+            api:
+                XcvrApi object
+            host_lane_count:
+                Integer, number of lanes on the host side
+            channel:
+                Integer, channel id of the group which the host lanes belong to (1 based)
+            speed:
+                Integer, the port speed of the host interface
+
+        Returns:
+            Integer, a mask of the active lanes on the host side
+            e.g. 0x5 for lane 0 and lane 2.
+        """
+        host_lane_mask = 0
+
+        if host_lane_count <= 0 or channel < 0 or speed <= 0:
+            self.log_error("Invalid input to get host lane mask - host_lane_count {} channel {} speed {}!".format(
+                            host_lane_count, channel, speed))
+            return host_lane_mask
+
+        if channel != 0:
+            appl = self.get_cmis_application_desired(api, host_lane_count, speed)
+            if appl < 1:
+                self.log_error("Failed to get host lanes as no suitable app {} found host_lane_count {} speed {}".format(
+                                appl, host_lane_count, speed))
+                return host_lane_mask
+
+            host_lane_assignment_option = api.get_host_lane_assignment_option(appl)
+            host_lane_start_bit = (host_lane_count * (channel - 1))
+            if host_lane_assignment_option & (1 << host_lane_start_bit):
+                host_lane_mask = ((1 << host_lane_count) - 1) << host_lane_start_bit
+        else:
+            host_lane_mask = (1 << host_lane_count) - 1
+
```

Comment:
can we add an error message if host_lane_mask is ZERO so that we know something went wrong?

Reply:
@mihirpat1 can we return (host_lane_mask, appl) so that caller can print the selected application and host_lane_mask for error and info

Reply:
I am now taking appl is an input parameter to this function as per our discussion.

Reply:
> @mihirpat1 can we test if the link is UP and Xcvrd crashes or PMON restarts, the link does not flap?

Yes, I have verified that link remains up on local as well as peer port when xcvrd crash/restart and PMON restart